### PR TITLE
[BEAM-2717] Emit coders in generated protos.

### PR DESCRIPTION
**Pull request by robertwb**
------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

**Comment by robertwb** [comment]
R: @charlesccychen

File: sdks/python/apache_beam/pipeline.py [hunk 1]
```diff
@@ -603,12 +605,36 @@ def visit_value(self, value, _):
...
+    # The RunnerAPI spec requires certain transforms to have KV inputs
+    # (and corresponding outputs).
+    # Currently we only upgrade to KV pairs.  If there is a need for more
+    # general shapes, potential conflicts will have to be resolved.
+    class ForceKvInputTypes(PipelineVisitor):
+      def enter_composite_transform(self, transform_node):
+        self.visit_transform(transform_node)
+
+      def visit_transform(self, transform_node):
+        if transform_node.transform and transform_node.transform.runner_api_requires_keyed_input():
```
Comment (charlesccychen) [inline sdks/python/apache_beam/pipeline.py]: Please fix line length and other presubmit issues.

File: sdks/python/apache_beam/pipeline.py [hunk 2]
```diff
@@ -603,12 +605,36 @@ def visit_value(self, value, _):
...
+          pcoll = transform_node.inputs[0]
+          pcoll.element_type = typehints.coerce_to_kv_type(
+              pcoll.element_type, transform_node.full_label)
+          if len(transform_node.outputs) == 1:
```
Comment (charlesccychen) [inline sdks/python/apache_beam/pipeline.py]: Can we comment that we only deal with the special case of one input / one output (is that sufficient in all scenarios?)

File: sdks/python/apache_beam/runners/pipeline_context.py [hunk 3]
```diff
@@ -124,6 +126,13 @@ def __init__(self, proto=None, default_environment=None):
           Environment(default_environment))
     else:
       self._default_environment_id = None
+    self.use_fake_coders = use_fake_coders
```
Comment (charlesccychen) [inline sdks/python/apache_beam/runners/pipeline_context.py]: Can we add a brief description of this feature, and if this is meant to be temporary, link to a JIRA to resolve this workaround?

File: sdks/python/apache_beam/runners/pipeline_context.py [hunk 4]
```diff
@@ -124,6 +126,13 @@ def __init__(self, proto=None, default_environment=None):
...
+
+  def coder_id_from_element_type(self, element_type):
```
Comment (charlesccychen) [inline sdks/python/apache_beam/runners/pipeline_context.py]: Please link to cleanup JIRA for once the workaround is no longer necessary.

File: sdks/python/apache_beam/pipeline.py (hunk 2 above)
Comment (robertwb) [reply sdks/python/apache_beam/pipeline.py]: Done.

File: sdks/python/apache_beam/runners/pipeline_context.py (hunk 4 above)
Comment (robertwb) [reply sdks/python/apache_beam/runners/pipeline_context.py]: Done.

File: sdks/python/apache_beam/runners/pipeline_context.py (hunk 3 above)
Comment (robertwb) [reply sdks/python/apache_beam/runners/pipeline_context.py]: Done.

**Comment by robertwb** [comment]
Resolved comments, precommits are happy. PTAL.

**Comment by charlesccychen** [comment]
Thanks! This LGTM.

**Pull request merged**